lsstseries/ensemble.py

```diff
@@ -7,17 +9,198 @@
 class ensemble:
     """ensemble object is a collection of light curve ids"""
 
-    def __init__(self, token=None):
+    def __init__(self, token=None, **kwargs):
         self.result = None  # holds the latest query
         self.token = token
+        self.data = None
 
-        self._time_col = "midPointTai"
-        self._flux_col = "psFlux"
-        self._err_col = "psFluxErr"
+        # Assign Default Values for critical column quantities
+        self._id_col = 'object_id'
+        self._time_col = 'midPointTai'
+        self._flux_col = 'psFlux'
+        self._err_col = 'psFluxErr'
+        self._band_col = 'band'
+
+        # Setup Dask Distributed Client
+        self.client = Client(**kwargs)  # arguments passed along to Client
+
+    def client_info(self):
+        """Calls the Dask Client, which returns cluster information
+
+        Parameters
+        ----------
+        None
+
+        Returns
+        ----------
+        self.client: `distributed.client.Client`
+            Dask Client information
+        """
+        return self.client  # Prints Dask dashboard to screen
+
+    def count(self, sort=True, ascending=False):
+        """Return the number of available rows/measurements for each lightcurve
+
+        Parameters
+        ----------
+        sort: `bool`, optional
+            Indicates whether the resulting counts should be sorted on counts
+        ascending: `bool`, optional
+            When sorting, indicates whether the result is ordering is ascending
+            or descending
+
+        Returns
+        ----------
+        counts: `pandas.series`
+            A series of counts by object
+        """
+        counts = self.data.groupby(self._id_col)[self._time_col].count().compute()
+        if sort:
+            return counts.sort_values(ascending=ascending)
+        else:
+            return counts
+
+    def dropna(self, threshold):
+        """wrapper for dask.dataframe.dropna
+
+        Parameters
+        ----------
+        threshold: `int`
+            The minimum number of nans present in a row needed to drop the row
+
+        Returns
+        ----------
+        ensemble: `lsstseries.ensemble.ensemble`
+            The ensemble object with nans removed according to the threshold
+            scheme
+        """
+        self.data = self.data[self.data.isnull().sum(axis=1) < threshold]
+        return self
+
+    def prune(self, threshold):
+        """remove objects with less observations than a given threshold
+
+        Parameters
+        ----------
+        threshold: `int`
+            The minimum number of observations needed to retain an object
+
+        Returns
+        ----------
+        ensemble: `lsstseries.ensemble.ensemble`
+            The ensemble object with pruned rows removed
+        """
+        counts = self.data.groupby(self._id_col).count()
```

Comment:
Originally, calling counts() was exactly what I did. I had a faster implementation that used counts to make a subset of ids that filtered the index. For some reason, Dask had a really hard time with this. Computing the subset list was making the workers unresponsive to future compute calls. Potentially because it was partially executing the task graph that was needed for downstream computes? There may be a quick solution, but I spent a while trying to find one with no luck. This should be something we revisit in the future, but the current implementation is one that operates with just a single compute() to play nice with Dask.